File: android/src/main/java/org/maplibre/reactnative/components/camera/MLRNCamera.java
package org.maplibre.reactnative.components.camera;

import android.content.Context;
import android.location.Location;
import android.util.Log;

import org.maplibre.android.camera.CameraPosition;
import org.maplibre.android.camera.CameraUpdate;
import org.maplibre.android.camera.CameraUpdateFactory;
import org.maplibre.android.geometry.LatLng;
import org.maplibre.android.geometry.LatLngBounds;
import org.maplibre.android.geometry.VisibleRegion;
import org.maplibre.android.location.OnCameraTrackingChangedListener;
import org.maplibre.android.location.modes.CameraMode;
import org.maplibre.android.maps.MapLibreMap;
import org.maplibre.android.maps.Style;

import org.maplibre.reactnative.components.AbstractMapFeature;
import org.maplibre.reactnative.components.location.LocationComponentManager;
import org.maplibre.reactnative.components.mapview.MLRNMapView;
import org.maplibre.reactnative.events.IEvent;
import org.maplibre.reactnative.events.MapUserTrackingModeEvent;
import org.maplibre.reactnative.events.MapChangeEvent;
import org.maplibre.reactnative.location.LocationManager;
import org.maplibre.reactnative.location.UserLocation;
import org.maplibre.reactnative.location.UserLocationVerticalAlignment;
import org.maplibre.reactnative.location.UserTrackingMode;
import org.maplibre.reactnative.location.UserTrackingState;
import org.maplibre.reactnative.utils.GeoJSONUtils;

import org.maplibre.reactnative.events.constants.EventTypes;

import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.WritableNativeMap;

import org.maplibre.geojson.Point;

import org.maplibre.android.location.permissions.PermissionsManager;

import androidx.annotation.NonNull;

public class MLRNCamera extends AbstractMapFeature {
    private MLRNCameraManager mManager;
    private MLRNMapView mMapView;

    private boolean hasSentFirstRegion = false;

    private CameraStop mDefaultStop;
    private CameraStop mCameraStop;
    private CameraUpdateQueue mCameraUpdateQueue;

    // private LocationComponent mLocationComponent;
    private LocationComponentManager mLocationComponentManager;

    private int mUserTrackingMode;
    private int mUserTrackingState = UserTrackingState.POSSIBLE;
    private int mUserLocationVerticalAlignment = UserLocationVerticalAlignment.CENTER;

    public static final int USER_LOCATION_CAMERA_MOVE_DURATION = 1000;

    private LocationManager mLocationManager;
    private UserLocation mUserLocation;

    private Point mCenterCoordinate;

    private boolean mAnimated;
    private double mHeading;
    private double mPitch;
    private double mZoomLevel = -1;

    private double mMinZoomLevel = -1;
    private double mMaxZoomLevel = -1;

    private LatLngBounds mMaxBounds;

    private boolean mFollowUserLocation;
    private String mFollowUserMode;

    private Context mContext;


    private LocationManager.OnUserLocationChange mLocationChangeListener = new LocationManager.OnUserLocationChange() {
        @Override
        public void onLocationChange(Location nextLocation) {
            if (getMapboxMap() == null || mLocationComponentManager == null || !mLocationComponentManager.hasLocationComponent() || (!mFollowUserLocation)) {
                return;
            }

            mUserLocation.setCurrentLocation(nextLocation);
            sendUserLocationUpdateEvent(nextLocation);
        }
    };

    private MapLibreMap.CancelableCallback mCameraCallback = new MapLibreMap.CancelableCallback() {
        @Override
        public void onCancel() {
            if (!hasSentFirstRegion) {
                mMapView.sendRegionChangeEvent(false);
                hasSentFirstRegion = true;
            }
        }

        @Override
        public void onFinish() {
            if (!hasSentFirstRegion) {
                mMapView.sendRegionChangeEvent(false);
                hasSentFirstRegion = true;
            }
        }
    };

    public MLRNCamera(Context context, MLRNCameraManager manager) {
        super(context);
        mContext = context;
        mManager = manager;
        mCameraUpdateQueue = new CameraUpdateQueue();

        mUserLocation = new UserLocation();
        mLocationManager = LocationManager.getInstance(context);
    }

    @Override
    public void addToMap(MLRNMapView mapView) {
        mMapView = mapView;

        setInitialCamera();
        updateMaxMinZoomLevel();
        updateMaxBounds();

        if (mCameraStop != null) {
            updateCamera();
        }

        if (mFollowUserLocation) {
            enableLocation();
        }
    }

    @Override
    public void removeFromMap(MLRNMapView mapView) {

    }

    public void setStop(CameraStop stop) {
        mCameraStop = stop;
        mCameraStop.setCallback(mCameraCallback);

        if (mMapView != null) {
            updateCamera();
        }
    }

    public void setDefaultStop(CameraStop stop) {
        mDefaultStop = stop;
    }

    public void setFollowPitch(double pitch) {
        mPitch = pitch;
        updateCameraPositionIfNeeded(true);
    }

    public void setMaxBounds(LatLngBounds bounds) {
        mMaxBounds = bounds;
        updateMaxBounds();
    }

    private void updateMaxBounds() {
        MapLibreMap map = getMapboxMap();
        if (map != null && mMaxBounds != null) {
            map.setLatLngBoundsForCameraTarget(mMaxBounds);
        }
    }

    private void updateMaxMinZoomLevel() {
        MapLibreMap map = getMapboxMap();
        if (map != null) {
            if (mMinZoomLevel >= 0.0) {
                map.setMinZoomPreference(mMinZoomLevel);
            }
            if (mMaxZoomLevel >= 0.0) {
                map.setMaxZoomPreference(mMaxZoomLevel);
            }
        }
    }

    private void setInitialCamera() {
        if (mDefaultStop != null) {
            mDefaultStop.setDuration(0);
            mDefaultStop.setMode(org.maplibre.reactnative.components.camera.constants.CameraMode.NONE);
            CameraUpdateItem item = mDefaultStop.toCameraUpdate(mMapView);
            item.run();
        }
    }

    private void updateCamera() {
        mCameraUpdateQueue.offer(mCameraStop);
        mCameraUpdateQueue.execute(mMapView);
    }

    private void updateUserTrackingMode(int userTrackingMode) {
        mUserLocation.setTrackingMode(userTrackingMode);
        IEvent event = new MapUserTrackingModeEvent(this, userTrackingMode);
        mManager.handleEvent(event);
    }

    private void updateUserLocation(boolean isAnimated) {
        if ((!mFollowUserLocation) || mUserLocation.getTrackingMode() == UserTrackingMode.NONE) {
            return;
        }

        if (mUserTrackingState == UserTrackingState.POSSIBLE) {
            updateUserLocationSignificantly(isAnimated);
        } else if (mUserTrackingState == UserTrackingState.CHANGED) {
            updateUserLocationIncrementally(isAnimated);
        }
    }

    private CameraPosition getUserLocationUpdateCameraPosition(double zoomLevel) {
        LatLng center = mUserLocation.getCoordinate();

        if (mUserLocationVerticalAlignment != UserLocationVerticalAlignment.CENTER) {
            VisibleRegion region = mMapView.getVisibleRegion(center, zoomLevel);

            switch (mUserLocationVerticalAlignment) {
                case UserLocationVerticalAlignment.TOP:
                    center = new LatLng(region.nearRight.getLatitude(), center.getLongitude());
                    break;
                case UserLocationVerticalAlignment.BOTTOM:
                    center = new LatLng(region.farLeft.getLatitude(), center.getLongitude());
                    break;
            }
        }

        return new CameraPosition.Builder().target(center).bearing(getDirectionForUserLocationUpdate()).tilt(mPitch).zoom(zoomLevel).build();
    }

    private double getDirectionForUserLocationUpdate() {
        // NOTE: The direction of this is used for map rotation only, not location layer rotation
        CameraPosition currentCamera = mMapView.getCameraPosition();
        double direction = currentCamera.bearing;

        int userTrackingMode = mUserLocation.getTrackingMode();
        if (userTrackingMode == UserTrackingMode.FollowWithHeading || userTrackingMode == UserTrackingMode.FollowWithCourse) {
            direction = mUserLocation.getBearing();
        } else if (mHeading != 0.0) {
            direction = mHeading;
        }

        return direction;
    }

    private void sendUserLocationUpdateEvent(Location location) {
        if (location == null) {
            return;
        }
        IEvent event = new MapChangeEvent(this, EventTypes.USER_LOCATION_UPDATED, makeLocationChangePayload(location));
        mManager.handleEvent(event);
    }

    private boolean hasSetCenterCoordinate() {
        CameraPosition cameraPosition = mMapView.getCameraPosition();
        LatLng center = cameraPosition.target;
        return center.getLatitude() != 0.0 && center.getLongitude() != 0.0;
    }

    static final double minimumZoomLevelForUserTracking = 10.5;
    static final double defaultZoomLevelForUserTracking = 14.0;

    private void updateUserLocationSignificantly(boolean isAnimated) {
        mUserTrackingState = UserTrackingState.BEGAN;

        double zoom = mZoomLevel;
        if (zoom < 0) {
            double cameraZoom = mMapView.getMapboxMap().getCameraPosition().zoom;
            if (cameraZoom < minimumZoomLevelForUserTracking) {
                zoom = defaultZoomLevelForUserTracking;
            } else {
                zoom = cameraZoom;
            }
        }
        CameraUpdate cameraUpdate = CameraUpdateFactory.newCameraPosition(getUserLocationUpdateCameraPosition(zoom));
        MapLibreMap.CancelableCallback cameraCallback = new MapLibreMap.CancelableCallback() {
            @Override
            public void onCancel() {
                mUserTrackingState = UserTrackingState.CHANGED;
            }

            @Override
            public void onFinish() {
                mUserTrackingState = UserTrackingState.CHANGED;
            }
        };

        if (isAnimated && hasSetCenterCoordinate()) {
            mMapView.animateCamera(cameraUpdate, cameraCallback);
        } else {
            mMapView.moveCamera(cameraUpdate, cameraCallback);
        }
    }

    private void updateUserLocationIncrementally(boolean isAnimated) {
        mUserTrackingState = UserTrackingState.BEGAN;

        CameraPosition cameraPosition = mMapView.getCameraPosition();
        CameraUpdate cameraUpdate = CameraUpdateFactory.newCameraPosition(getUserLocationUpdateCameraPosition(cameraPosition.zoom));

        MapLibreMap.CancelableCallback callback = new MapLibreMap.CancelableCallback() {
            @Override
            public void onCancel() {
                mUserTrackingState = UserTrackingState.CHANGED;
            }

            @Override
            public void onFinish() {
                mUserTrackingState = UserTrackingState.CHANGED;
            }
        };

        if (isAnimated) {
            mMapView.easeCamera(cameraUpdate, USER_LOCATION_CAMERA_MOVE_DURATION, true, callback);
        } else {
            mMapView.moveCamera(cameraUpdate, callback);
        }
    }

    private void enableLocation() {
        if (!PermissionsManager.areLocationPermissionsGranted(mContext)) {
            return;
        }

        if (!mLocationManager.isActive()) {
            mLocationManager.enable();
        }

        mMapView.getMapboxMap().getStyle(new Style.OnStyleLoaded() {
            @Override
            public void onStyleLoaded(@NonNull Style style) {
                enableLocationComponent(style);
            }
        });
    }

    private void enableLocationComponent(@NonNull Style style) {
        updateUserLocation(false);
        updateLocationLayer(style);

        Location lastKnownLocation = mLocationManager.getLastKnownLocation();
        mLocationManager.addLocationListener(mLocationChangeListener);

        if (lastKnownLocation != null) {
            mLocationChangeListener.onLocationChange(lastKnownLocation);

            postDelayed(new Runnable() {
                @Override
                public void run() {
                    mMapView.sendRegionDidChangeEvent();
                }
            }, 200);
        }

    }

    private void updateLocationLayer(@NonNull Style style) {
        if (mLocationComponentManager == null) {
            mLocationComponentManager = mMapView.getLocationComponentManager();
        }

        mLocationComponentManager.update(style);
        mLocationComponentManager.setFollowUserLocation(mFollowUserLocation);

        if (mFollowUserLocation) {
            mLocationComponentManager.setCameraMode(UserTrackingMode.getCameraMode(mUserTrackingMode));
            mLocationComponentManager.addOnCameraTrackingChangedListener(new OnCameraTrackingChangedListener() {
                @Override
                public void onCameraTrackingChanged(int currentMode) {
                    int userTrackingMode;

                    switch (currentMode) {
                        case CameraMode.NONE:
                            userTrackingMode = UserTrackingMode.NONE;
                            break;
                        case CameraMode.TRACKING:
                            userTrackingMode = UserTrackingMode.FOLLOW;
                            break;
                        case CameraMode.TRACKING_COMPASS:
                            userTrackingMode = UserTrackingMode.FollowWithHeading;
                            break;
                        case CameraMode.TRACKING_GPS:
                            userTrackingMode = UserTrackingMode.FollowWithCourse;
                            break;
                        default:
                            userTrackingMode = UserTrackingMode.NONE;
                    }

                    updateUserTrackingMode(userTrackingMode);
                }

                @Override
                public void onCameraTrackingDismissed() {
                }
            });
        } else {
            mLocationComponentManager.setCameraMode(CameraMode.NONE);
        }
    }

    public void setMinZoomLevel(double zoomLevel) {
        mMinZoomLevel = zoomLevel;
        updateMaxMinZoomLevel();
    }

    public void setMaxZoomLevel(double zoomLevel) {
        mMaxZoomLevel = zoomLevel;
        updateMaxMinZoomLevel();
    }

    public void setZoomLevel(double zoomLevel) {
        mZoomLevel = zoomLevel;
        updateCameraPositionIfNeeded(false);
    }

    private CameraPosition buildCamera(CameraPosition previousPosition, boolean shouldUpdateTarget) {
        CameraPosition.Builder builder = new CameraPosition.Builder(previousPosition).bearing(mHeading).tilt(mPitch).zoom(mZoomLevel);

        if (shouldUpdateTarget) {
            builder.target(GeoJSONUtils.toLatLng(mCenterCoordinate));
        }

        return builder.build();
    }

    private void updateCameraPositionIfNeeded(boolean shouldUpdateTarget) {
        if (mMapView != null) {
            CameraPosition prevPosition = mMapView.getCameraPosition();
            CameraUpdate cameraUpdate = CameraUpdateFactory.newCameraPosition(buildCamera(prevPosition, shouldUpdateTarget));

            if (mAnimated) {
                mMapView.easeCamera(cameraUpdate);
            } else {
                mMapView.moveCamera(cameraUpdate);
            }
        }
    }

    public void setUserTrackingMode(int userTrackingMode) {
        int oldTrackingMode = mUserTrackingMode;
        mUserTrackingMode = userTrackingMode;
        updateUserTrackingMode(userTrackingMode);

        switch (mUserTrackingMode) {
            case UserTrackingMode.NONE:
                mUserTrackingState = UserTrackingState.POSSIBLE;
                break;
            case UserTrackingMode.FOLLOW:
            case UserTrackingMode.FollowWithCourse:
            case UserTrackingMode.FollowWithHeading:
                if (oldTrackingMode == UserTrackingMode.NONE) {
                    mUserTrackingState = UserTrackingState.POSSIBLE;
                }
                break;

        }

        if (getMapboxMap() != null) {
            updateLocationLayer(getMapboxMap().getStyle());
        }
    }


    public void setFollowUserLocation(boolean value) {
        mFollowUserLocation = value;
        updatedFollowUserMode();
    }

    public void setFollowUserMode(String mode) {
        mFollowUserMode = mode;
        updatedFollowUserMode();
    }

    private void updatedFollowUserMode() {
        if (mFollowUserLocation) {
            setUserTrackingMode(UserTrackingMode.fromString(mFollowUserMode));
        } else {
            setUserTrackingMode(UserTrackingMode.NONE);
        }
    }

    MapLibreMap getMapboxMap() {
        if (mMapView == null) {
            return null;
        }
        return mMapView.getMapboxMap();
    }

    /**
     * Create a payload of the location data per the web api geolocation spec
     * https://dev.w3.org/geo/api/spec-source.html#position
     *
     * @return
     */
    private WritableMap makeLocationChangePayload(Location location) {
        WritableMap positionProperties = new WritableNativeMap();
        WritableMap coords = new WritableNativeMap();

        coords.putDouble("longitude", location.getLongitude());
        coords.putDouble("latitude", location.getLatitude());
        coords.putDouble("altitude", location.getAltitude());
        coords.putDouble("accuracy", location.getAccuracy());
        // A better solution will be to pull the heading from the compass engine, 
        // unfortunately the api is not publicly available in the mapbox sdk
        coords.putDouble("heading", location.getBearing());
        coords.putDouble("course", location.getBearing());
        coords.putDouble("speed", location.getSpeed());

        positionProperties.putMap("coords", coords);
        positionProperties.putDouble("timestamp", location.getTime());
        return positionProperties;
    }
}

File: android/src/main/java/org/maplibre/reactnative/components/camera/MLRNCamera.kt
package org.maplibre.reactnative.components.camera

import android.Manifest
import android.content.Context
import android.location.Location
import android.util.AttributeSet
import androidx.annotation.RequiresPermission
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReactContext
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableMap
import com.facebook.react.uimanager.UIManagerHelper
import com.facebook.react.uimanager.events.Event
import com.facebook.react.uimanager.events.EventDispatcher
import org.maplibre.android.camera.CameraPosition
import org.maplibre.android.camera.CameraUpdateFactory.newCameraPosition
import org.maplibre.android.geometry.LatLngBounds
import org.maplibre.android.location.OnCameraTrackingChangedListener
import org.maplibre.android.location.permissions.PermissionsManager
import org.maplibre.android.maps.MapLibreMap
import org.maplibre.android.maps.MapLibreMap.CancelableCallback
import org.maplibre.android.maps.Style
import org.maplibre.reactnative.components.AbstractMapFeature
import org.maplibre.reactnative.components.camera.constants.CameraEasing
import org.maplibre.reactnative.components.location.LocationComponentManager
import org.maplibre.reactnative.components.mapview.MLRNMapView
import org.maplibre.reactnative.location.LocationManager
import org.maplibre.reactnative.location.LocationManager.OnUserLocationChange
import org.maplibre.reactnative.location.TrackUserLocationMode
import org.maplibre.reactnative.location.TrackUserLocationState
import org.maplibre.reactnative.location.UserLocation

class MLRNCamera(
    context: Context,
) : AbstractMapFeature(
        context,
    ) {
    @Suppress("UNUSED_PARAMETER")
    constructor(context: Context, attrs: AttributeSet?) : this(context)

    @Suppress("UNUSED_PARAMETER")
    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : this(context)

    private var mapView: MLRNMapView? = null
    private var hasSentFirstRegion = false
    private val updateQueue = CameraUpdateQueue()

    private var stop: CameraStop? = null

    private var initialViewState: CameraStop? = null

    private var minZoom: Double? = null
    private var maxZoom: Double? = null

    private var maxBounds: LatLngBounds? = null

    private var trackUserLocation = TrackUserLocationMode.NONE
    private var userTrackingState = TrackUserLocationState.POSSIBLE

    private var locationComponentManager: LocationComponentManager? = null

    private val locationManager: LocationManager = LocationManager.getInstance(context)
    private val userLocation = UserLocation()

    private val locationChangeListener: OnUserLocationChange =
        object : OnUserLocationChange {
            override fun onLocationChange(location: Location) {
                if (mapView!!.mapLibreMap == null ||
                    locationComponentManager == null ||
                    !locationComponentManager!!.hasLocationComponent() ||
                    trackUserLocation == TrackUserLocationMode.NONE
                ) {
                    return
                }

                userLocation.setCurrentLocation(location)
            }
        }

    val eventDispatcher: EventDispatcher?
        get() {
            val reactContext = context as ReactContext

            return UIManagerHelper.getEventDispatcherForReactTag(reactContext, id)
        }

    val surfaceId: Int
        get() {
            val reactContext = context as ReactContext

            return UIManagerHelper.getSurfaceId(reactContext)
        }

    private val cameraCallback: CancelableCallback =
        object : CancelableCallback {
            override fun onCancel() {
                if (!hasSentFirstRegion) {
                    mapView!!.sendRegionChangeEvent(false)
                    hasSentFirstRegion = true
                }
            }

            override fun onFinish() {
                if (!hasSentFirstRegion) {
                    mapView!!.sendRegionChangeEvent(false)
                    hasSentFirstRegion = true
                }
            }
        }

    @RequiresPermission(allOf = [Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION])
    override fun addToMap(mapView: MLRNMapView) {
        this@MLRNCamera.mapView = mapView

        setInitialCamera()
        updateMaxMinZoom()
        updateMaxBounds()

        if (stop != null) {
            updateCamera()
        }

        if (trackUserLocation != TrackUserLocationMode.NONE) {
            enableLocation()
        }
    }

    override fun removeFromMap(mapView: MLRNMapView) {}

    fun handleImperativeStop(stop: ReadableMap?) {
        val mapView = this.mapView ?: return
        val stop =
            CameraStop.fromReadableMap(context, stop ?: Arguments.createMap(), cameraCallback)

        updateQueue.offer(stop)
        updateQueue.execute(mapView)
    }

    fun setStop(stop: ReadableMap?) {
        this.stop =
            CameraStop.fromReadableMap(context, stop ?: Arguments.createMap(), cameraCallback)

        if (mapView != null) {
            updateCamera()
        }
    }

    fun setInitialViewState(stop: CameraStop?) {
        initialViewState = stop
    }

    fun hasInitialViewState(): Boolean = initialViewState != null

    fun setMaxBounds(bounds: LatLngBounds?) {
        maxBounds = bounds
        updateMaxBounds()
    }

    private fun updateMaxMinZoom() {
        val map = maplibreMap
        if (map != null) {
            val minZoom = minZoom
            if (minZoom != null) {
                map.setMinZoomPreference(minZoom)
            }

            val maxZoom = maxZoom
            if (maxZoom != null) {
                map.setMaxZoomPreference(maxZoom)
            }
        }
    }

    private fun updateMaxBounds() {
        val map = maplibreMap
        if (map != null && maxBounds != null) {
            map.setLatLngBoundsForCameraTarget(maxBounds)
        }
    }

    private fun setInitialCamera() {
        if (initialViewState != null) {
            initialViewState!!.duration = 0
            initialViewState!!.easing = CameraEasing.NONE
            val item = initialViewState!!.toCameraUpdate(mapView!!)

            item.run()
        }
    }

    private fun updateCamera() {
        updateQueue.offer(stop)
        updateQueue.execute(mapView!!)
    }

    private fun updateUserTrackingMode(trackUserLocationMode: Int) {
        userLocation.trackingMode = trackUserLocationMode

        val event = OnTrackUserLocationChangeEvent(surfaceId, id, trackUserLocationMode)
        eventDispatcher?.dispatchEvent(event)
    }

    private fun updateUserLocation() {
        if (trackUserLocation == TrackUserLocationMode.NONE || userLocation.trackingMode == TrackUserLocationMode.NONE) {
            return
        }

        if (userTrackingState == TrackUserLocationState.POSSIBLE) {
            updateUserLocationSignificantly()
        } else if (userTrackingState == TrackUserLocationState.CHANGED) {
            updateUserLocationIncrementally()
        }
    }

    private fun getUserLocationUpdateCameraPosition(zoomLevel: Double): CameraPosition {
        val center = userLocation.coordinate

        return CameraPosition
            .Builder()
            .target(center)
            .bearing(
                directionForUserLocationUpdate,
            ).tilt(stop?.pitch ?: 0.0)
            .zoom(zoomLevel)
            .build()
    }

    private val directionForUserLocationUpdate: Double
        get() {
            // NOTE: The direction of this is used for map rotation only, not location layer rotation
            val currentCamera = mapView!!.cameraPosition
            var direction = currentCamera.bearing

            val userTrackingMode = userLocation.trackingMode
            if (userTrackingMode == TrackUserLocationMode.HEADING || userTrackingMode == TrackUserLocationMode.COURSE) {
                direction = userLocation.bearing
            } else if (stop?.bearing != null) {
                direction = stop?.bearing ?: direction
            }

            return direction
        }

    private fun updateUserLocationSignificantly() {
        userTrackingState = TrackUserLocationState.BEGAN

        val cameraUpdate =
            newCameraPosition(getUserLocationUpdateCameraPosition(mapView!!.mapLibreMap!!.cameraPosition.zoom))
        val cameraCallback: CancelableCallback =
            object : CancelableCallback {
                override fun onCancel() {
                    userTrackingState = TrackUserLocationState.CHANGED
                }

                override fun onFinish() {
                    userTrackingState = TrackUserLocationState.CHANGED
                }
            }

        mapView!!.moveCamera(cameraUpdate, cameraCallback)
    }

    private fun updateUserLocationIncrementally() {
        userTrackingState = TrackUserLocationState.BEGAN

        val cameraPosition = mapView!!.cameraPosition
        val cameraUpdate =
            newCameraPosition(getUserLocationUpdateCameraPosition(cameraPosition.zoom))

        val callback: CancelableCallback =
            object : CancelableCallback {
                override fun onCancel() {
                    userTrackingState = TrackUserLocationState.CHANGED
                }

                override fun onFinish() {
                    userTrackingState = TrackUserLocationState.CHANGED
                }
            }

        mapView!!.moveCamera(cameraUpdate, callback)
    }

    @RequiresPermission(allOf = [Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION])
    private fun enableLocation() {
        if (!PermissionsManager.areLocationPermissionsGranted(context)) {
            return
        }

        if (!locationManager.isActive()) {
            locationManager.enable()
        }

        mapView!!.mapLibreMap!!.getStyle { style ->
            enableLocationComponent(
                style,
            )
        }
    }

    private fun enableLocationComponent(style: Style) {
        updateUserLocation()
        updateLocationLayer(style)

        val lastKnownLocation = locationManager.lastKnownLocation
        locationManager.addLocationListener(locationChangeListener)

        if (lastKnownLocation != null) {
            locationChangeListener.onLocationChange(lastKnownLocation)

            postDelayed({ mapView!!.sendRegionDidChangeEvent() }, 200)
        }
    }

    private fun updateLocationLayer(style: Style) {
        if (locationComponentManager == null) {
            locationComponentManager = mapView!!.locationComponentManager
        }

        locationComponentManager!!.update(style)
        locationComponentManager!!.setFollowUserLocation(trackUserLocation != TrackUserLocationMode.NONE)

        if (trackUserLocation != TrackUserLocationMode.NONE) {
            locationComponentManager!!.setCameraMode(
                TrackUserLocationMode.getCameraMode(
                    trackUserLocation,
                ),
            )
            locationComponentManager!!.addOnCameraTrackingChangedListener(
                object :
                    OnCameraTrackingChangedListener {
                    override fun onCameraTrackingChanged(currentMode: Int) {
                        val userTrackingMode =
                            when (currentMode) {
                                org.maplibre.android.location.modes.CameraMode.NONE -> TrackUserLocationMode.NONE
                                org.maplibre.android.location.modes.CameraMode.TRACKING -> TrackUserLocationMode.DEFAULT
                                org.maplibre.android.location.modes.CameraMode.TRACKING_COMPASS -> TrackUserLocationMode.HEADING
                                org.maplibre.android.location.modes.CameraMode.TRACKING_GPS -> TrackUserLocationMode.COURSE
                                else -> TrackUserLocationMode.NONE
                            }

                        updateUserTrackingMode(userTrackingMode)
                    }

                    override fun onCameraTrackingDismissed() {
                    }
                },
            )
        } else {
            locationComponentManager!!.setCameraMode(org.maplibre.android.location.modes.CameraMode.NONE)
        }
    }

    fun setMinZoomLevel(zoomLevel: Double) {
        minZoom = zoomLevel
        updateMaxMinZoom()
    }

    fun setMaxZoomLevel(zoomLevel: Double) {
        maxZoom = zoomLevel
        updateMaxMinZoom()
    }

    fun setTrackUserLocation(trackUserLocationMode: Int) {
        val oldTrackingMode = trackUserLocation
        trackUserLocation = trackUserLocationMode
        updateUserTrackingMode(trackUserLocationMode)

        when (trackUserLocation) {
            TrackUserLocationMode.NONE -> {
                userTrackingState = TrackUserLocationState.POSSIBLE
            }

            TrackUserLocationMode.DEFAULT, TrackUserLocationMode.COURSE, TrackUserLocationMode.HEADING -> {
                if (oldTrackingMode ==
                    TrackUserLocationMode.NONE
                ) {
                    userTrackingState = TrackUserLocationState.POSSIBLE
                }
            }
        }

        if (maplibreMap != null) {
            updateLocationLayer(maplibreMap!!.style!!)
        }
    }

    private val maplibreMap: MapLibreMap?
        get() {
            if (mapView == null) {
                return null
            }
            return mapView!!.mapLibreMap
        }

    inner class OnTrackUserLocationChangeEvent(
        surfaceId: Int,
        viewId: Int,
        private val trackUserLocationMode: Int,
    ) : Event<OnTrackUserLocationChangeEvent>(surfaceId, viewId) {
        override fun getEventName() = "onTrackUserLocationChange"

        override fun getEventData(): WritableMap =
            Arguments.createMap().apply {
                putString(
                    "trackUserLocation",
                    TrackUserLocationMode.toString(trackUserLocationMode),
                )
            }
    }
}
